apps/wappush/js/wappush.js

```diff
@@ -282,49 +266,127 @@
   }
 
   /**
-   * Removes the notifications for messages that have been deleted
+   * Reacts to a new message being added to the database by showing a
+   * notification or directly showing the message if required by the action
+   * field.
    *
-   * @param {Object} message The deleted message.
+   * @param {Object} obj The new message in untyped form (i.e. JSON)
    */
-  function wpm_onMessageDeleted(message) {
-    wpm_clearNotifications(+message.timestamp);
+  function wpm_onNew(obj) {
+    DUMP('A new message added to the DB');
+
+    var message = new ParsedMessage(obj);
+
+    if (message.action === 'signal-high' ||
+        message.action === 'execute-high')
+    {
+      /* We just decrease the number of pending messages instead of invoking
+       * wpm_finish() since we don't want to start the close procedure. */
+      pendingMessages--;
+      wpm_displayWapPushMessage(message.getUniqueId());
+    } else {
+      wpm_sendNotification(message).then(() => wpm_finish());
+    }
+  }
+
+  /**
+   * Reacts to a message being updated in the database by updating the
+   * notification and possibly the message being shown to the user.
+   *
+   * @param {Object} obj The new message in untyped form (i.e. JSON)
+   */
+  function wpm_onUpdate(obj) {
+    DUMP('A message was updated in the DB');
+
+    var message = new ParsedMessage(obj);
+
+    // Update the message if we're currently displaying it
+    if (message.getUniqueId() === displayedSiMessage.getUniqueId()) {
+      displayedSiMessage = message;
+      SiSlScreenHelper.populateScreen(message);
+    }
+
+    // This updates the existing notification
+    wpm_sendNotification(message).then(() => wpm_finish());
+  }
+
+  /**
+   * Removes the notifications for messages that have been deleted and closes
+   * the app if the message being viewed has been deleted.
+   *
+   * @param {Object} obj The deleted message in untyped form (i.e. JSON)
+   */
+  function wpm_onDelete(obj) {
+    DUMP('A message was deleted');
+
+    var message = new ParsedMessage(obj);
+
+    // Close the application if we're displaying the deleted message
+    if (displayedSiMessage &&
+      displayedSiMessage.getUniqueId() === message.getUniqueId()) {
```

Comment:
nit: I'd indent this by 2 more spaces (just below the previous line's `displayedSiMessage`) so that it's crystal-clear that it's part of the condition and not a line code